code-styling/bash/README.md

```diff
@@ -0,0 +1,423 @@
+
+- When splitting, a line MUST be indented and begin with a pipe '|' or logical 
+  compound ('||' and '&&').  Example:
+  
+  ```bash
+  command \
+      | pipe \
+      && other
+  ```
+
+- The `; do` and `; then` keywords MUST be on the same line as the `while`, `for` 
+  or `if` it belongs to.
+
+<!-- ======================================================================= -->
+
+## Declaration 
+
+### Arrays
+
+- For arrays, either use `"$@"` or `$*` MUST be used .
+  An unquoted `$@` MUST NOT be used.
+
+### Checks
+
+- The new BASH double square bracket builtin test operator (`[[ ... ]]`) MUST be
+  used. The old single square bracket test operator `[` or explicit call
+  to `test` MUST NOT be used.
+
+- To avoid confusion about what is being tested for `-z` or `-n` MUST be used.
+  Use `[[ -z $variable ]]`  or `[[ -n $variable ]]` rather than just `[[ $variable ]]`.
+
+- A single equal sign SHOULD be used when checking values. Double or triple 
+  signs SHOULD NOT be used.
+  Example: `[[ "${NAME}" = "John" ]]` 
+
+- When referring to the current file `${BASH_SOURCE[0]}` SHOULD be used, even if 
+  it is sourced from a parent script. In other cases, `${0}` SHOULD be used.
+
+- As the `errexit` flag is used, `|| true` MUST be appended to commands that 
+  might cause an error which should not halt the script. (For instance when the
+  command output is assigned to a variable).
+
+### Command options
+
+- Long options MUST be used if they are available, 
+  for instance, use  `logger --priority --verbose` instead of `logger -p -v`. 
+
+### Command substitution
+
+- Sub-shell syntax `$(command)` MUST be used for command substitution rather
+  than backticks  <code>\`command`</code>.
+
+### Functions
+
+- Except for function packaging code (see below) all code MUST go in a function. 
+  Even if it's just one `main` function.
+
+- Functions MUST be defined without the `function` keyword, with round brackets, 
+  like this: `myfunc() { ... }`
+
+- All functions MUST be located together in the file just below constants. 
+  Executable code MUST NOT exists between functions.
+
+- Complex one-liners MUST be placed in standalone functions with an appropriate 
+  name and description.
+
+#### Function packaging
+
+- If it should be possible to call certain code both directly _and_ sourced from 
+  other scripts, it MUST be wrapped in a "self" check, like this: 
+  
+  ```bash
+  if [ "${BASH_SOURCE[0]}" == "${0}" ]; then
+      # direct call to file
+      main "${@}"
+  fi  # else file is included from another script
+  ```
+
+### Strings
+
+- For strings, single quotes `'` MUST be used when substitution is not required.
+- For strings, double quotes `"` MUST be used when substitution is required.
+
+### User messaging
+
+- User output SHOULD be formatted to follow the Heroku output format[^2](#references).
+- Warnings and errors MUST go to `STDERR`, anything than could be parsed SHOULD go to `STDOUT`.
```

Comment:
Only for self generated output, not for command called from this script, because other binaries may not follow this guide